### PR TITLE
Initial Retrie plugin

### LINKED ISSUE #103: Retrie plugin
http://hackage.haskell.org/package/retrie 

Retrie is a modern alternative to HaRe, and possibly better maintained as it is used by Facebook.

It can be used to implement refactorings like:

- Inline functions with `--unfold`
- Extract functions, with `--fold`
- Inline and extract type synonyms, with `--type-forward` and `--type-backward`

It could be also used to implement renames and moves, but those would require custom scripts.

### PR DESCRIPTION
**Pull request by pepeiborra**
Supports RULES, functions and type synonyms

Future work:

- Handling names properly (when https://github.com/facebookincubator/retrie/issues/10 is fixed)
- Suggestions for pattern synonyms (when https://github.com/facebookincubator/retrie/issues/15 is released)
- Refactorings: rename, extract, move, etc..
- Automatically add imports when unfolding
- Proper support for workspace folders

Closes #103
 
/cc @xich

### CODE REVIEW

File: src/Ide/Plugin.hs [hunk 1]
```diff
@@ -334,6 +335,14 @@ makeExecuteCommands ecs lf ide = do
 
 -- | Runs a plugin command given a PluginId, CommandId and
 -- arguments in the form of a JSON object.
+wrapUnhandledExceptions ::
```
Comment (maksbotan) [inline src/Ide/Plugin.hs:338]: Hmm, the haddock is now misplaced.

Comment (pepeiborra) [reply src/Ide/Plugin.hs:335]: Fixed, thanks!

File: README.md [hunk 2]
```diff
@@ -71,6 +71,10 @@ This is *very* early stage software.
 
   This will cause compilation errors if a dependency contains invalid Haddock markup, though in a future version of GHC (hopefully 8.12), [these will be demoted to warnings](https://gitlab.haskell.org/ghc/ghc/-/merge_requests/2377).
 
+ - Integration with [retrie](https://hackage.haskell.org/package/retrie)
+
+   ![Retrie](https://i.imgur.com/VFKx1He.gif)
```
Comment (puffnfresh) [inline README.md:76]: I get a 404 for this image.

Comment (pepeiborra) [reply README.md]: Fixed now, thanks!

File: include/ghc-api-version.h [hunk 3]
```diff
@@ -0,0 +1,10 @@
+#ifndef GHC_API_VERSION_H
+#define GHC_API_VERSION_H
+
+#ifdef GHC_LIB
+#define MIN_GHC_API_VERSION(x,y,z) MIN_VERSION_ghc_lib(x,y,z)
```
Comment (bubba) [inline include/ghc-api-version.h:5]: Does this mean we can build hls with -fghc-lib now? That might solve the hlint ghc-lib mismatch issue @jneira

Comment (pepeiborra) [reply include/ghc-api-version.h:5]: I don't know about that, I have never tried to build ghcide with ghc-lib. Ping @cocreature

**Comment by pepeiborra** [comment]
> LGTM
> 
> My only concern is that it provides code actions for the whole file, rather than in the context.
> 
> I wonder if we shouldn't split this into diagnostics and code actions, much as hlint (in hie) does.

The intent is for the code actions to be attached to the relevant lines, not to the whole file. Is that not what you see?

I don't follow your second comment, what diagnostics?

**Comment by alanz** [comment]
> The intent is for the code actions to be attached to the relevant lines, not to the whole file. Is that not what you see?

TBH, I have not run the code, just looked at it.  And I see now that the `range` is converted to a `pos` and then used as a filter.

> I don't follow your second comment, what diagnostics?

The normal process is that a plugin provides diagnostics, which show up as markings in the file and a list of issues to address.  When the cursor is placed on a line with a diagnostic, this is returned to the code action provider which can propose fixes.

It seems to me that you will not know that retrie *has* a rewrite suggestion unless the cursor is on a line with a rewrite in it?

If so, my suggestion is to run retrie in an unconstrained way, on the whole file, and turn the suggestions into diagnostics. This is how we process hlint, it runs to generate diagnostics about possible changes, and if you go to the line you get an action to actually change it.

**Comment by pepeiborra** [comment]
> If so, my suggestion is to run retrie in an unconstrained way, on the whole file, and turn the suggestions into diagnostics. This is how we process hlint, it runs to generate diagnostics about possible changes, and if you go to the line you get an action to actually change it.

Retrie does not generate suggestions. It takes code modding commands as inputs, e.g. "unfold foo" and applies them.

**Comment by alanz** [comment]
Ok, any reason not to merge then?

**Comment by pepeiborra** [comment]
> Ok, any reason not to merge then?

Merge away, I'll open issues for the future work items and any TODOs left in the code.

**Pull request merged**